src/task/executor.rs

```diff
@@ -70,91 +71,147 @@ impl Clone for Kicker {
+        let task = unsafe { Arc::from_raw(id as *mut SpawnTask) };
+        if !task.mark_notified() {
+            mem::forget(task);
+            return;
+        }
+
+        if let Some(UnfinishedWork(w)) = self.push_work(UnfinishedWork(task.clone())) {
+            match task.kicker.kick(Box::new(CallTag::Spawn(w))) {
+                // If the queue is shutdown, then the tag will be notified
+                // eventually. So just skip here.
+                Err(Error::QueueShutdown) => (),
+                Err(e) => panic!("unexpected error when canceling call: {:?}", e),
+                _ => (),
+            }
         }
+        mem::forget(task);
+    }
+
+    fn clone_id(&self, id: usize) -> usize {
+        let task = unsafe { Arc::from_raw(id as *mut SpawnTask) };
+        let t = task.clone();
+        mem::forget(task);
+        Arc::into_raw(t) as usize
     }
 
-    pub fn resolve(self, success: bool) {
-        // it should always be canceled for now.
-        assert!(success);
-        poll(&Arc::new(self.clone()), true);
+    fn drop_id(&self, id: usize) {
+        unsafe { Arc::from_raw(id as *mut SpawnTask) };
     }
 }
 
-impl Notify for SpawnNotify {
-    fn notify(&self, _: usize) {
-        if thread::current().id() == self.worker_id {
-            poll(&Arc::new(self.clone()), false)
-        } else {
-            let mut ctx = self.ctx.lock();
-            if ctx.kicked {
-                return;
-            }
-            ctx.notify(Box::new(CallTag::Spawn(self.clone())));
-            ctx.kicked = true;
-        }
+pub struct UnfinishedWork(Arc<SpawnTask>);
+
+impl UnfinishedWork {
+    pub fn finish(self, cq: &CompletionQueue) {
+        resolve(cq, self.0, true);
     }
 }
 
 /// Poll the future.
 ///
 /// `woken` indicates that if the cq is kicked by itself.
-fn poll(notify: &Arc<SpawnNotify>, woken: bool) {
-    let mut handle = notify.handle.lock();
-    if woken {
-        notify.ctx.lock().kicked = false;
-    }
-    if handle.is_none() {
-        // it's resolved, no need to poll again.
-        return;
-    }
-    match handle.as_mut().unwrap().poll_future_notify(notify, 0) {
-        Err(_) | Ok(Async::Ready(_)) => {
-            // Future stores notify, and notify contains future,
-            // hence circular reference. Take the future to break it.
-            handle.take();
+fn poll(cq: &CompletionQueue, task: Arc<SpawnTask>, woken: bool) {
+    let mut init_state = if woken { NOTIFIED } else { IDLE };
+    // TODO: maybe we need to break the loop to avoid hunger.
+    loop {
+        match task
+            .state
+            .compare_exchange(init_state, POLLING, Ordering::SeqCst, Ordering::Acquire)
+        {
+            Ok(_) => {}
+            Err(COMPLETED) => return,
+            Err(s) => panic!("unexpected state {}", s),
+        }
+
+        let id = &*task as *const SpawnTask as usize;
+
+        match unsafe { &mut *task.handle.get() }
+            .as_mut()
```

Comment:
Why call `.as_mut` on the `&mut` reference? I think it should be called on the `*mut` if handle might be null or not used at all, if handle is never null.

Reply:
Because the type of handle is `UnsafeCell<SpawnHandle>`, and `SpawnHandle` is an alias of `Option<Spawn<BoxFuture<(), ()>>>`. `as_mut` is the method of `Option`.